### ISSUE TITLE
[Project / Organisation] Milestone achievements

### ISSUE THREAD

**Issue by rubyFeedback** [libui-ng/libui-ng#45]
I think if you look at the original libui then you may see andlabs was most active from 2015 (and perhaps before that, not sure when he wrote the Go code) to about 2019 or so, give or take.

libui-ng already made some improvements - see the changelog here by cody:

https://github.com/libui-ng/libui-ng/blob/master/CHANGELOG.md

For many "downstream" projects, in particular those who maintain bindings to other languages, I think it may be useful
to have some kind of "milestone" achievement "flag" set. That is sort of like saying "ok, tons of improvements were made to
libui-ng, it is time to transition into it and polish up our bindings".

The rationale here, respectively the use case, may be that you can get the downstream projects to update their own bindings
to libui (libui-ng rather). This is, I believe, one reason why the CHANGELOG is good to have. Perhaps the format could group a bit between API changes and other changes, but that's besides the point here.

At some later point in time I assume libui-ng may also become less active. That is in the nature of projects that they tend to have some higher activity and then lateron it subsides again normally.

My suggestion here would be to set some reasonable milestone, something that can be achieved and that may be realistic. I don't know when or how - this is up to the devs to decide (those who know C actually). But once that first milestone is set, it would be nice if there could be a recommendation by libui-ng to say "ok guys, this is now the first milestone, we reached it in 6 months and polished things. Now you folks can improve the libui-bindings that you maintain". I think this may help with adoption of libui-ng too, at the least based on the improvements and patch set made.

If you look at the stars of libui it has 10.1k; libui-ng right now has 34 so many people still don't know it. I believe this may change as time passes by, and for that case it would be nice to know at which point the first "recommendation" is done by upstream, whoever maintains this. I know that kojix2 (who wrote the ffi wrapper code for libui) is considering this and Andy from the glimmer DSL probably would appreciate this as well considering the bug reports he made to the original libui.

Note that with "milestone" I mostly refer to something that is feasible, e. g. some small bug fixes, documentation improvements, some patches that are to be applied and so forth. It's a bit arbitrary to say "at this point milestone 1.0 has been reached", but I am thinking here mostly in terms of libui-bindings maintainers here, if that makes any sense what I wrote here so far. It's just a suggestion anyway - please feel free to ignore/modify as it fits. Open source projects often have to "pass the baton" so to speak.

**Comment by cody271** [comment]
> My suggestion here would be to set some reasonable milestone, something that can be achieved and that may be realistic. I don't know when or how - this is up to the devs to decide (those who know C actually). But once that first milestone is set
>
That's an easy question to answer: The first milestone is [merging in the old PRs](https://github.com/libui-ng/libui-ng/issues/8) from the original project.

I'm sorry if this wasn't already super clear.

**Comment by cody271** [comment]
> Perhaps the format could group a bit between API changes and other changes, but that's besides the point here.
>
Yup, the changelog format definitely needs to be improved.

**Comment by msink** [comment]
> The first milestone is merging in the old PRs from the original project.

Well, many of that PRs are actually unfinished - for merging some feature it have to be implemented for Linux+macOS+Windows, but many PRs have implementation only for one or two systems.

So, by "merging" you actually mean "finish and merge", or what?

**Comment by cody271** [comment]
> 
> So, by "merging" you actually mean "finish and merge", or what?
>
Yes exactly, "finish and merge" for some of them.

**Comment by szanni** [comment]
My take on things would be for the first milestone

- merge all bug fix PRs
- merge all _reasonable_ PRs (there are a number I would personally not merge, as they IMO are out of scope)
- _try_ to avoid breaking the existing API for the time being
- release first libui-ng alpha

This way is will be fairly straight forward for libui user to migrate to libui-ng. Get all the fixes and enhancements with minimal hassle.

Then, for subsequent milestones, pull in API breaking things. There are a few things (especially in the table code) that I believe will need API breaking changes. Some things are currently fairly messy, due to the complexity that was very difficult to foresee.

Just my 2 cents.

**Comment by cody271** [comment]
> My take on things would be for the first milestone
>

Looks right, this is pretty much the plan so far. 👍

**Comment by cody271** [comment]
> * merge all _reasonable_ PRs (there are a number I would personally not merge, as they IMO are out of scope)

I've already marked some on the master table with `Out of Scope`. Which did I miss?

**Comment by AndyObtiva** [comment]
Good news!!!

@cody271

My [talk proposal](https://andymaleh.blogspot.com/2022/09/rubyconf-2022-building-native-gui-apps.html) for [RubyConf 2022](https://rubyconf.org/) titled "Building Native GUI Apps in Ruby" has been accepted! It will be given sometime between Nov 29 and Dec 1, 2022 in Houston, TX, USA.

Abstract:

> Ruby is an excellent choice for building desktop apps with a native GUI (Graphical User Interface) that looks familiar on Mac, Windows, and Linux. In fact, Ruby pushes the boundaries of developing such apps in brand new ways not seen in web development by supporting very lightweight and declarative GUI syntax including bidirectional data-binding, thanks to [Glimmer DSL for LibUI](https://github.com/AndyObtiva/glimmer-dsl-libui), a gem that [won a Fukuoka Ruby 2022 Special Award](https://andymaleh.blogspot.com/2022/02/glimmer-dsl-for-libui-wins-fukuoka-ruby.html). In this talk, I will cover concepts like the GUI DSL, data-binding, custom controls, area graphics, drag & drop, MVC/MVP pattern, and scaffolding, with sample demos.

Now, I noticed that you have not made any releases in libui-ng yet.

Wouldn't it be great if you made a release just before I gave the talk and I got to show it off to the whole wide world!?!!!!!

If you could make a release a month in advance of the conference at least (or preferably much earlier), like by Halloween (Oct 31, 2022) at the latest, that would give @kojix2 and I enough time to add some or all of the necessary updates to use it in [Glimmer DSL for LibUI](https://github.com/AndyObtiva/glimmer-dsl-libui) so that I could present it at [RubyConf 2022](https://rubyconf.org/).

Otherwise, I stand by the recommendation I made to @andlabs in [libui](https://github.com/andlabs/libui/issues/549) to shift the release plan into micro-releases that are mini-complete-features, each released with a different minor/patch version number (e.g. 1.1, 1.2, 1.3, or 1.1.2, 1.1.3, 1.1.4, or 1.1.2.0, 1.1.3.0, 1.1.4.0, whatever feels best), to enable consumers like @kojix2, other binding developers, and I to incrementally add the released features to our projects instead of getting surprised by huge bulks of work that delay us from releasing for several months.

Here is what I told @andlabs in [this issue](https://github.com/andlabs/libui/issues/549) ("&lt;NOTE&gt; Active fork `libui-ng`"):

> @andlabs It's good to hear back from you. One recommendation I have to beat time management issues is to simply split releases into much smaller chunks. For example, if you add support for displaying an image under an area for Windows, Mac, and Linux, then release right away as alpha5 (or alpha 4.2). Afterwards, if you add support for triggering events when clicking table columns (e.g. for sorting) on Windows, Mac, and Linux, again, release immediately as a very tiny alpha6 (or alpha 4.3) release. Sure, you will end up with a lot of mini releases, but this will ensure that any efforts you undertake are very small and managable no matter your work schedule, and you also ensure that consumers of the project continue to receive value from the project via small releases even if you end up with very little time to work on the project. This is the process I followed while building [Glimmer DSL for LibUI](https://github.com/AndyObtiva/glimmer-dsl-libui) in Ruby by the way. I documented my process as the [Glimmer Process](https://github.com/AndyObtiva/glimmer/blob/master/PROCESS.md), which is inspired by Toyota lean manufacturing processes, like [Kanban](https://en.wikipedia.org/wiki/Kanban). In any case, I look forward to your continued support of libui. [Glimmer DSL for LibUI](https://github.com/AndyObtiva/glimmer-dsl-libui) won a [Fukuoka Ruby 2022 Special Award](https://www.digitalfukuoka.jp/topics/187?locale=ja) from the creator of Ruby himself and the Fukuoka Ruby committee as they were truly impressed by libui in addition to the [Glimmer DSL for LibUI](https://github.com/AndyObtiva/glimmer-dsl-libui) high-level Ruby binding. I'm sure they'd appreciate libui's continued progress going forward, in addition to other consumers of the project who would benefit from its continued support too!

After the upcoming release (which I hope happens much sooner than later, and definitely before Halloween 2022), I would really look forward to seeing micro-releases of mini-complete-features of libui-ng.

Cheers and thank you for your continued effort in maintaining libui inside the libui-ng fork!

**Comment by kojix2** [comment]
@AndyObtiva
Hi there, I see you're going to be speaking at RubyConf. Congratulations.
Do you think I should make a pre-release gem and release it to the public?

@cody271
For you libui-ng guys, which shared libraries should I include in the gem?
I see that you have a variety of files placed here. Can I get the files you put here and re-distribute them?
https://github.com/libui-ng/libui-ng/actions/runs/2849713683

M1 mac needs a different build, right?

**Comment by AndyObtiva** [comment]
I have a confident feeling that libui-ng is ready for a real release, with many more releases to follow, so I would rather just support a real release. I’d rather not put effort into an unsure prerelease. In the brave new world of open-source, developers should feel empowered to release early and often as they could always patch up any discovered bugs quickly if needed just like they do in the GTK project. So, there is no excuse for having very lengthy long term releases like old Microsoft software lifecycles from the 90’s. It’s important to make quick releases to cut the total release risk down and get finished work out to get used and tested by real software engineers in the wild for real feedback. This prevents the possibility of a project stalling forever and eventually dying like what happened in the original libui project for a while. 

@cody271 could you please make a release with everything new that is ready for use on all platforms, given the ability to make more follow up releases when needed in the future?

**Comment by szanni** [comment]
> My [talk proposal](https://andymaleh.blogspot.com/2022/09/rubyconf-2022-building-native-gui-apps.html) for [RubyConf 2022](https://rubyconf.org/) titled "Building Native GUI Apps in Ruby" has been accepted! It will be given sometime between Nov 29 and Dec 1, 2022 in Houston, TX, USA.

Great!

> Now, I noticed that you have not made any releases in libui-ng yet.

That is correct. The first release was supposed to have all libui PRs completed and merged (within reasonable scope).

> If you could make a release a month in advance of the conference at least, like by Halloween (Oct 31, 2022) at the latest (or preferably much earlier), that would give @kojix2 and I enough time to add some or all of the necessary updates to use it in [Glimmer DSL for LibUI](https://github.com/AndyObtiva/glimmer-dsl-libui) so that I could present it at [RubyConf 

The API should remain backwards compatible for the first 1-2 releases (including things like `const`) to make transitioning for developers and language bindings as easy as possible.

> Otherwise, I stand by the recommendation I made to @andlabs in [libui](https://github.com/andlabs/libui/issues/549) to shift the release plan into micro-releases that are mini-complete-features, each released with a different minor/patch version number (e.g. 1.1, 1.2, 1.3, or 1.1.2, 1.1.3, 1.1.4, or 1.1.2.0, 1.1.3.0, 1.1.4.0, whatever feels best), to enable consumers like @kojix2, other binding developers, and I to incrementally add the released features to our projects instead of getting surprised by huge bulks of work that delay us from releasing for several months.

libui-ng will be using semver, so `0.0.0`, `0.1.0`, `0.1.1`, etc.

> @cody271 I want to make a pre-release gem that includes libui-ng. For you libui-ng guys, which shared libraries should I include in the gem? I see that you have a variety of files placed here. Can I get the files you put here and re-distribute them? https://github.com/libui-ng/libui-ng/actions/runs/2849713683

I am not familiar with ruby gems and possible restrictions, but I would probably recommend using the static libraries, quite possibly with #113 as libc/++ might be a problem because of #67.

> M1 mac needs a different build, right?

Mac libraries should be fat (`x86_64` and `arm64`), so M1 should be supported by default.



> I have a confident feeling that libui-ng is ready for a real release

I am personally not quite sure, as there are quite a few things that are still very broken.

> [...] I would rather just support a real release. I’d rather not put effort into an unsure prerelease. In the brave new world of open-source, developers should feel empowered to release early and often as they could always patch up any discovered bugs quickly if needed just like they do in the GTK project.

A good point, but to be honest, having to work with GTK is quite frustrating right now. It is frankly quite amazing how broken GTK is with such popularity.

I would personally prefer more testing around the existing components. I discovered quite a few bugs with the introductions of some automated and manual test. I'm sadly pretty slammed with other commitments right now. I am however not completely opposed releases in it's current form, just a little hesitant.

**Comment by AndyObtiva** [comment]
A user just asked me about missing functionality in libui that is very basic, and yet it is in the upcoming change log already: uiOpenFolder

Here is my conversation with that user:

> Tom ten Thij @tomtt Sep 28 10:19
> Question: the open_file window only works to open a file. Is there some way to allow it to open a directory instead?
> 
> Andy Maleh @AndyObtiva Sep 28 11:26
> Yes, that would be open_folder, but it won't be available until the next libui-ng C library release, which should happen hopefully in October. Once that release happens, Glimmer DSL for LibUI will get updated to add open_folder. Otherwise, today, you could use Glimmer DSL for Tk with the Tk.chooseDirectory call to open a directory. Alternatively, you could use Glimmer DSL for SWT in JRuby, which supports directory_dialog.open (as demonstrated in Hello, Directory Dialog!)

It would be great if we could please that user by making an incremental alpha release. I’m not asking for perfection. But, if the current libui-ng has less bugs than the last libui release, then it is definitely releasable. 

Is that too much to ask?

I mean if I’m forced, I could use git diff releases in a halfhazard approach (the Ruby libui binding that I depend on in Glimmer DSL for LibUI already supports that), but I doubt that would give users the confidence of an official release. 

Still, if releasing an alpha incrementally is too much of a problem for you, I can adapt to the situation if I have to, but I thought I’d double check if a release is possible ASAP first.

**Comment by cody271** [comment]
> > I have a confident feeling that libui-ng is ready for a real release
> 
> I am personally not quite sure, as there are quite a few things that are still very broken.
>
Yes I would just reiterate in agreement here with all previous comments above from @szanni, we're still quite a ways off from our goals for the first release!

> I mean if I’m forced, I could use git diff releases in a halfhazard approach (the Ruby libui binding that I depend on in Glimmer DSL for LibUI already supports that), but I doubt that would give users the confidence of an official release.
>
IMHO this is exactly what you should do for now: pin your dependency to a known working version of the master branch.
For example it would just be "libui-ng version [b0d9d3b](https://github.com/libui-ng/libui-ng/commit/b0d9d3b744a5aff6fdbe7ba89ca882bf3c21f4f1)" as of today.

And if for some reason this _really_ doesn't work for your use case, I guess I am open to tagging a `0.0.0` prerelease just for the language bindings.. but I sure would like to avoid that, if possible.


> > My [talk proposal](https://andymaleh.blogspot.com/2022/09/rubyconf-2022-building-native-gui-apps.html) for [RubyConf 2022](https://rubyconf.org/) titled "Building Native GUI Apps in Ruby" has been accepted! It will be given sometime between Nov 29 and Dec 1, 2022 in Houston, TX, USA.
> 
> Great!
>
This is indeed very cool news! Let us know how else we can help.

**Comment by cody271** [comment]
> For you libui-ng guys, which shared libraries should I include in the gem? I see that you have a variety of files placed here. Can I get the files you put here and re-distribute them? https://github.com/libui-ng/libui-ng/actions/runs/2849713683
> 
> M1 mac needs a different build, right?

Yes please continue to use the [nightly.link](https://nightly.link/libui-ng/libui-ng/workflows/build/master) integration for redistributable CI builds for now.
Mac builds are still labelled as `macOS-x64` there, but the binaries are actually fat builds for both architectures:

```
libui-ng$ file build/meson-out/libui.dylib
build/meson-out/libui.dylib: Mach-O universal binary with 2 architectures:
Mach-O 64-bit dynamically linked shared library x86_64
Mach-O 64-bit dynamically linked shared library arm64
```
